Documentation/doc-dataframe.org

```diff
@@ -0,0 +1,192 @@
+
+Column names are strings and cell values may be of any type. Currently
+defined operations on data frames are: creation, copying, adding rows
+and columns, creating a data frame slice, selecting a single element
+and mapping over whole date frame, its rows and over row columns.
+
+* Creating data frames
+
+** New data frames
+
+To create a data frame call the function ~make-data-frame~. The first
+argument is a sequence containing column names and the rest are rows
+(each row must be a sequence of the same length as the first
+argument). Column names must be unique strings.
+
+#+BEGIN_SRC common-lisp
+  (defparameter *data-frame*
+    (make-data-frame '("name"             "address"         "number")
+                     ;;---------------------------------------------------
+                     '("Chris Richardson" "Franz, Inc."     "510-548-3600")
+                     '("Colin Meldrum"    "Franz, Inc."     "510-548-3600")
+                     '("Scott McKay"      "Symbolics, Inc." "617-221-1000")
+                     '("Bill York"        "Lucid, Inc."     "415-xxx-yyyy")
+                     '("Paul Weineke"     "Lucid, Inc."     "415-xxx-yyyy")))
+#+END_SRC
+
+** Adding rows and columns
+
+It is possible to add new rows and columns to an existing data
+frame. To do that we use destructive operators ~add-rows!~ and
+~add-cols!~. The value type is not constrained by the library. Let's
+add few extra rows.
+
+#+BEGIN_SRC common-lisp
+  (add-rows! *data-frame*
+             '("Daniel Kochmański" "TurtleWare" "516-xxx-yyy")
+             '("Anna Durke"        "TurtleWare" "516-xxx-yyy")
+             '("Emmy Grahams"      "TurtleWare" "516-xxx-yyy")
+             '("Caroline Tarsky"   "Just, Inc." :invalid)
+             '("Andrew Smith"      "Just, Inc." :hidden)
+             '("David Smith"       "Just, Inc." "666-xxx-yyy")
+             '("Emily Bublinski"   "Doit"       "333-xxx-yyy"))
+#+END_SRC
+
+Adding columns is a little more tricky, because to add a column we
+specify a function of the entire row. The function result is used as
+the value for a new column in given row.
+
+#+BEGIN_SRC common-lisp
+  (add-cols! *data-frame*
+             "Name length"
+             (lambda (row-index row)
+               (declare (ignore row-index))
+               (length (ref *data-frame* row "name")))
+             "Country"
+             (lambda (row-index row)
+               (declare (ignore row-index))
+               (let ((address (ref *data-frame* row "address")))
+                 (switch (address :test (lambda (elt pat)
+                                          (member elt pat :test #'string=)))
+                   ('("Franz, Inc." "Symbolics, Inc." "Lucid, Inc.") "USA")
+                   ('("TurtleWare" "Just, Inc.")                     "Poland")
+                   (otherwise                                        "Unknown")))))
+#+END_SRC
+
+Notice that we've used a function ~ref~. We could have used ~elt~
```

Comment:
But shouldn't we prefer `ref`, because using `elt` exposes the implementation? (this applies here and in the code snippet here and in doc-dataframe.lisp).

Reply:
I've removed the paragraph.